grove/plugin/range_req_handler.go

```diff
@@ -0,0 +1,225 @@
+	return &cfg
+}
+
+// rangeReqHandlerOnRequest determines if there is a Range header, and puts the ranges in *d.Context as a []byteRanges
+func rangeReqHandlerOnRequest(icfg interface{}, d OnRequestData) bool {
+	rHeader := d.R.Header.Get("Range")
+	if rHeader == "" {
+		log.Debugf("No Range header found")
+		return false
+	}
+	log.Debugf("Range string is: %s", rHeader)
+	// put the ranges [] in the context so we can use it later
+	byteRanges := parseRangeHeader(rHeader)
+	*d.Context = byteRanges
+	return false
+}
+
+func rangeReqHandleBeforeCacheLookup(icfg interface{}, d BeforeCacheLookUpData, overRideFunc func(string)) {
+	cfg, ok := icfg.(*rangeRequestConfig)
+	if !ok {
+		log.Errorf("range_req_handler config '%v' type '%T' expected *rangeRequestConfig\n", icfg, icfg)
+		return
+	}
+	if cfg.Mode == "store_ranges" {
+		sep := "?"
+		if strings.Contains(d.DefaultCacheKey, "?") {
+			sep = "&"
+		}
+		newKey := d.DefaultCacheKey + sep + "grove_range_req_handler_plugin_data=" + d.Req.Header.Get("Range")
+		overRideFunc(newKey)
+		log.Debugf("range_req_handler: store_ranges default key:%s, new key:%s", d.DefaultCacheKey, newKey)
+	}
+}
+
+// rangeReqHandleBeforeParent changes the parent request if needed (mode == get_full_serve_range)
+func rangeReqHandleBeforeParent(icfg interface{}, d BeforeParentRequestData) {
+	log.Debugf("rangeReqHandleBeforeParent calling.")
+	rHeader := d.Req.Header.Get("Range")
+	if rHeader == "" {
+		log.Debugf("No Range header found")
+		return
+	}
+	log.Debugf("Range string is: %s", rHeader)
+	cfg, ok := icfg.(*rangeRequestConfig)
+	if !ok {
+		log.Errorf("range_req_handler config '%v' type '%T' expected *rangeRequestConfig\n", icfg, icfg)
+		return
+	}
+	if cfg.Mode == "get_full_serve_range" {
+		// get_full_serve_range means get the whole thing from parent/org, but serve the requested range. Just remove the Range header from the upstream request
+		d.Req.Header.Del("Range")
+	}
+	return
+}
+
+// rangeReqHandleBeforeRespond builds the 206 response
+// Assume all the needed ranges have been put in cache before, which is the truth for "get_full_serve_range" mode which gets the whole object into cache.
+func rangeReqHandleBeforeRespond(icfg interface{}, d BeforeRespondData) {
+	log.Debugf("rangeReqHandleBeforeRespond calling\n")
+	ictx := d.Context
+	ctx, ok := (*ictx).([]byteRange)
+	if !ok {
+		log.Errorf("Invalid context: %v", ictx)
+	}
+	if len(ctx) == 0 {
+		return // there was no (valid) range header
+	}
+
+	cfg, ok := icfg.(*rangeRequestConfig)
+	if !ok {
+		log.Errorf("range_req_handler config '%v' type '%T' expected *rangeRequestConfig\n", icfg, icfg)
+		return
+	}
+	if cfg.Mode == "store_ranges" {
+		return // no need to do anything here.
+	}
+
+	multipartBoundaryString := ""
+	originalContentType := d.Hdr.Get("Content-type")
+	*d.Hdr = web.CopyHeader(*d.Hdr) // copy the headers, we don't want to mod the cacheObj
+	if len(ctx) > 1 {
+		//multipart = true
+		multipartBoundaryBytes := make([]byte, 8)
```

Comment:
I think this should be 16 bytes, a UUID. 8 bytes might not be enough to be globally unique, and might appear in the body, breaking everything. Also, generating a random number is expensive, and this is in the request path, where performance is critical. Can you generate the UUID once, and reuse it? Should be able to do that with a `startup` hook, and set it in the context.